### PR TITLE
Add Repo Contents API endpoints.

### PR DESCRIPTION
**Pull request by scott-fleischman**
This adds the create, update and delete file endpoints from https://developer.github.com/v3/repos/contents/

The code in the PR works great.

I would say it's about 80% done in terms of the coding guidelines. Still to be done are updating the contents example from the samples (which is also out of date with respect to the current master branch). I wanted to wait to update those until I got approval on the current code structure.

I am opening the Pull Request now to get feedback on some of the code organizational choices and naming choices. I am happy to change them to whatever you think best. I am not particularly attached to my choices so far--I mainly picked what seemed to be a reasonably good start.

In particular I added a new module called `GitHub.Endpoints.Repos.Contents`. This mirrors the structure of the GitHub API.

I also moved `contentFor` and related functions to the new module, since they are located there in the GitHub API.

I added the new endpoints to the same module.

I added new datatypes to the module `GitHub.Data.Content` because that seemed like the best fit so far.

### CODE REVIEW

**Comment by scott-fleischman** [comment]
One further design consideration: the content fields for `CreateFile` and `UpdateFIle` are required to be in base64 encoding.

Right now there is nothing in this API that clues you in to that fact--you have to read the API documentation.

Should we do one of the following?
1. Change the name of the field to something like `contentBase64`
2. Automatically do the base64 encoding (perhaps using the `base64-bytestring` package).

**Comment by scott-fleischman** [comment]
@phadej do you have any initial feedback on this PR?

File: src/GitHub/Data/Content.hs [hunk 1]
```diff
@@ -5,6 +5,9 @@
 --
 module GitHub.Data.Content where
 
+import Data.Aeson.Types        (KeyValue)
```
Comment (phadej) [inline src/GitHub/Data/Content.hs]: we'll need to bump lower bump on aeson for this

File: src/GitHub/Data/Content.hs [hunk 2]
```diff
@@ -87,3 +155,74 @@ instance FromJSON ContentInfo where
...
+instance ToJSON Author where
+  toJSON (Author { authorName   = name
+                 , authorEmail  = email
+                 }) = object
+                 [ "name"      .= name
+                 , "email"     .= email
+                 ]
+
+instance ToJSON CreateFile where
+  toJSON (CreateFile { createFilePath       = path
```
Comment (phadej) [inline src/GitHub/Data/Content.hs]: I'd use `RecordWildCards`, this is extremely ugly. (No need to touch old code though)

File: src/GitHub/Data/Content.hs [hunk 3]
```diff
@@ -87,3 +155,74 @@ instance FromJSON ContentInfo where
...
+                     }) = object $
+                     [ "path"              .= path
+                     , "message"           .= message
+                     , "sha"               .= sha
+                     ]
+                     ++ "branch"           .=? branch
+                     ++ "author"           .=? author
+                     ++ "committer"        .=? committer
+
+(.=?) :: ToJSON v => KeyValue t => Text -> Maybe v -> [t]
```
Comment (phadej) [inline src/GitHub/Data/Content.hs]: cool!

Comment (phadej) [inline src/GitHub/Data/Content.hs]: With exolicit `[Aeson.Pair]` this will be compatible with old aesons, though we can bump lower bump too.

**Comment by scott-fleischman** [comment]
@phadej I updated the code based on your comments.

I also updated the Contents sample code so that it builds. I added it to the cabal file so it builds as part of github-samples. Finally, I added an example using the new create/update/delete file code.

Let me know if you would like anything else done on this PR.

**Comment by phadej** [comment]
Great, thanks!

**Pull request merged**